### PR TITLE
Refactor ADU PoC with new API design

### PR DESCRIPTION
**Pull request by ewertons**
## Purpose
Update the PoC code with the latest version of the API design (i.e., have no "ADU client").

## Does this introduce a breaking change?

```
[x] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

### CODE REVIEW

File: libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c [hunk 1]
```diff
@@ -148,459 +116,264 @@ static bool prvAreUpdateIdsEqual(
...
-                &xDeviceInformation,
-                ( int32_t ) pxAduClient->xState,
-                &pxAduClient->xUpdateRequest.workflow,
-                ( pxAduClient->pxLastInstallResult != NULL ) ? &xLastInstallResult : NULL,
-                xPropertiesPayload,
-                &xPropertiesPayload ) ) )
+    return eAzureIoTSuccess;
+}
+
+AzureIoTResult_t AzureIoTADUClient_GetResponse( AzureIoTHubClient_t * pxAzureIoTHubClient,
```
Comment (danewalton) [inline libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c:256]: was this the API which we had decided to be a `SendResponse()`?

File: libs/azure-iot-middleware-freertos/source/include/azure_iot_adu_client.h [hunk 2]
```diff
@@ -237,109 +237,128 @@ typedef enum AzureIoTADUState
...
+{
+    uint8_t * pucProvider;
+    uint32_t ulProviderLength;
+    uint8_t * pucName;
+    uint32_t ulNameLength;
+    uint8_t * pucVersion;
+    uint32_t ulVersionLength;
+} AzureIoTADUUpdateId_t;
+
+typedef struct AzureIoTADUCompatility
```
Comment (danewalton) [inline libs/azure-iot-middleware-freertos/source/include/azure_iot_adu_client.h:266]:
```suggestion
typedef struct AzureIoTADUCompatibility
```

File: libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c (hunk 1 above)
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c]: Yes. It's now changed to the SendResponse pattern we decided for.

File: libs/azure-iot-middleware-freertos/source/include/azure_iot_adu_client.h [hunk 3]
```diff
@@ -237,109 +237,128 @@ typedef enum AzureIoTADUState
...
+{
+    uint8_t * pucDeviceManufacturer;
+    uint32_t ulDeviceManufacturerLength;
+    uint8_t * pucDeviceModel;
+    uint32_t ulDeviceModelLength;
+} AzureIoTADUCompatility_t;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/source/include/azure_iot_adu_client.h:267]: Fixed.

File: libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c [hunk 4]
```diff
@@ -148,459 +116,264 @@ static bool prvAreUpdateIdsEqual(
...
-    return xResult;
-}
-
-AzureIoTADUUpdateStepState_t AzureIoTADUClient_GetState( AzureIoTADUClient_t * pxAduClient )
-{
-    return pxAduClient->xUpdateStepState;
-}
+    /*Send state update about the stage of the ADU process we are in */
+    return eAzureIoTSuccess;
+}
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c:407]: Fixed.

File: libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c [hunk 5]
```diff
@@ -148,459 +116,264 @@ static bool prvAreUpdateIdsEqual(
...
+                                                 AzureIoTHubClientADUDeviceInformation_t * pxDeviceInformation,
+                                                 AzureIoTADUUpdateRequest_t * pxAduUpdateRequest,
+                                                 AzureIoTADUState_t xAgentState,
+                                                 uint8_t * pucBuffer,
+                                                 uint32_t ulBufferSize,
+                                                 uint32_t * pulRequestId )
+{
+    az_iot_adu_ota_device_information xBaseDeviceInformation; // TODO: fill up.
+    az_iot_adu_ota_workflow xBaseWorkflow; // TODO: fill up.
+    az_iot_adu_ota_install_result xInstallResult; // TODO: fill up.
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/source/azure_iot_adu_client.c]: Not anymore, I addressed these.

File: demos/sample_azure_iot_adu/sample_azure_iot_adu.c [hunk 6]
```diff
@@ -341,19 +344,199 @@ static uint32_t prvSetupNetworkCredentials( NetworkCredentials_t * pxNetworkCred
...
+            xAzureIoTAduOtaUpdateRequest.pxFileUrls[ 0 ].pucUrl,
+            xAzureIoTAduOtaUpdateRequest.pxFileUrls[ 0 ].ulUrlLength ),
+        &xUrlHost, &xUrlPath );
+
+    /* TODO: remove this hack. */
+    char pcNullTerminatedHost[ 128 ];
+    ( void ) memcpy( pcNullTerminatedHost, az_span_ptr( xUrlHost ), az_span_size( xUrlHost ) );
+    pcNullTerminatedHost[ az_span_size( xUrlHost ) ] = '\0';
+
+    xResult = prvConnectHTTP( &xHTTPTransport, ( const char * ) pcNullTerminatedHost );
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/sample_azure_iot_adu/sample_azure_iot_adu.c:424]: That's a very good idea. This is one of those fixes that I would include in the final cleanup, which we planned for. But the earlier we fix these the better.

[the remark this answers is not in the document]
Comment (ewertons) [reply demos/sample_azure_iot_adu/sample_azure_iot_adu.c:424]: Done as you suggested.

**Review by danewalton — APPROVED** [review]
just need the code style script to be run over these

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c [hunk 7]
```diff
@@ -125,19 +103,43 @@ static az_span generate_update_id_string(
...
+static az_span get_json_writer_remaining_buffer(az_span original_buffer, az_json_writer* jw)
+{
+    return az_span_slice_to_end(
+        original_buffer, az_span_size(az_json_writer_get_bytes_used_in_destination(jw)));
+}
+
+static az_result generate_step_id(az_span buffer, uint32_t step_index, az_span* step_id)
+{
+    az_result result;
+    *step_id = buffer;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:132]: Yes, to save the original pointer address.

File: demos/sample_azure_iot_adu/sample_azure_iot_adu.c [hunk 8]
```diff
@@ -338,22 +341,243 @@ static uint32_t prvSetupNetworkCredentials( NetworkCredentials_t * pxNetworkCred
...
+    *pxPath = az_span_slice_to_end( xUrl, lPathPosition );
+}
+
+static AzureIoTResult_t prvDownloadUpdateImageIntoFlash()
+{
+    AzureIoTResult_t xResult;
+    AzureIoTHTTPResult_t xHttpResult;
+    AzureIoTHTTP_t xHTTP;
+    char * pucHttpDataBufferPtr;
+    uint32_t pulHttpDataLength;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/sample_azure_iot_adu/sample_azure_iot_adu.c:376]: Changed to ulHttpDataBufferLength. Thanks.

File: demos/sample_azure_iot_adu/sample_azure_iot_adu.c [hunk 9]
```diff
@@ -175,6 +164,20 @@ static const AzureIoTHubClientADUDeviceInformation_t xADUDeviceInformation =
...
+static uint8_t ucScratchBuffer[ 512 ];
+
+/* Command buffers */
+static uint8_t ucCommandResponsePayloadBuffer[ 256 ];
+
+/* Reported Properties buffers */
+static uint8_t ucReportedPropertiesUpdate[ 1500 ];
+static uint32_t ulReportedPropertiesUpdateLength;
+
+uint8_t ucAduContextBuffer[ ADU_CONTEXT_BUFFER_SIZE ];
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/sample_azure_iot_adu/sample_azure_iot_adu.c:179]: Absolutely. Do you know how many times I got that feedback? :)

File: libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (danewalton) [reply libs/azure-iot-middleware-freertos/libraries/azure-sdk-for-c/sdk/src/azure/iot/az_iot_adu_ota.c:132]: ahh didn't see it used on line 138

File: demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c [hunk 10]
```diff
@@ -175,19 +237,43 @@ void vHandleWritableProperties( AzureIoTHubClientPropertiesResponse_t * pxMessag
...
+                                &xAzureIoTHubClient,
                                 &xJsonReader,
+                                &xAzureIoTAduOtaUpdateRequest,
+                                ucAduContextBuffer,
+                                ADU_CONTEXT_BUFFER_SIZE );
+
+            if ( xAzIoTResult != eAzureIoTSuccess )
+            {
+                LogError( ( "AzureIoTADUClient_ParseRequest failed: result 0x%08x", xAzIoTResult ) );
+                * pulWritablePropertyResponseBufferLength = 0;
```
[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:254]: Interestingly, I ran the script. It did not catch this.
I fixed it manually.

[the remark this answers is not in the document]
Comment (ewertons) [reply demos/projects/ESPRESSIF/aduotaupdate/main/azure_iot_freertos_esp32_pnp_model.c:254]: Here and other places.

**Pull request merged**